REST API as documented on https://app.swaggerhub.com/apis/microprofile/starter/1

Karm opened the pull request:
Hello, the REST API is finished.

For the time being it is available on  https://test-start.microprofile.io/api/1.0.0 but next time something is merged into master, master will overwrite that deployment.

You can play with it locally on localhost:8080/api/1.0.0/ as

Logging into DynamoDB also works:
See:
```
$ aws  --profile cesar --region us-west-1 dynamodb scan --table-name microprofile_test_starter_log  | grep traffic -A2
            "trafficSource": {
                "S": "REST"
            },
--
            "trafficSource": {
                "S": "WEB"
            },

```

# Changes:

    * Updates logging do DynamoDB to log traffic source, WEB or REST
    * Makes JVM shutdown on OOM ExitOnOutOfMemoryError (just as a precaution)
    * REST API Examples: see [REST-README.md](https://github.com/Karm/microprofile-starter/blob/rest/src/main/resources/REST-README.md)
    * Adds Version string based on git HEAD to both Web UI and REST API
    * README.md file references REST API REST-README.md documentation
    * Thorntail stays at 2.2.1
    * Maven exec plugin uses version.txt.sh to generate version.txt that is then available via Version bean
    * Both REST at /api/1.0.0 and Web UI display the version number (either a tag or tag+commits and hash)
    * API Application and Endpoint separated into 2 classes
    * Error json messages contain error code
    * ETag generation takes Version into account
    * REST-README.md adds links to ETag spec
    * REST-README.md adds examples for Windows users

# REST API Examples:

  see  see [REST-README.md](https://github.com/Karm/microprofile-starter/blob/rest/src/main/resources/REST-README.md)

# UI change: Version number
![vesion-MicroProfile](https://user-images.githubusercontent.com/691097/58874522-5a1e6280-86c9-11e9-9318-49ac9083bef6.png)

Signed-off-by: Michal Karm Babacek <karm@fedoraproject.org>

Karm commented:
Fixes #91

rdebusscher commented on src/main/java/org/eclipse/microprofile/starter/rest/API.java:
```diff
@@ -0,0 +1,77 @@
+ */
+@ApplicationPath("/api")
+@Path("/1.0.0")
+public class API extends Application {
```
Uncommon to define a JAX-RS resource within the Application class. Separate this into 2 classes.

Karm replied:
Fixed.

rdebusscher commented on src/main/resources/REST-README.md:
````diff
@@ -0,0 +1,190 @@
+# Get available MicroProfile versions
+
+```
+$ curl https://start.microprofile.io/api/1.0.0/mpVersion
````
We should mention an alternative for Windows Users

Karm replied:
Chapter for Windows users added see [Windows users](https://github.com/Karm/microprofile-starter/blob/rest/src/main/resources/REST-README.md#windows-users)

Emily-Jiang replied:
can we make the url to be shorter: https://start.microprofile.io/mpVersion
as I think we should always use the latest version of MP Starter?

Karm replied:
@Emily-Jiang I would like to leave the ```/api``` there so as it indicates what it is. 
It is true that we can move the API version from URL param to request header and offer the latest by default I guess... 
The most straightforward thing to do now would be to change "1.0.0" to "1" and leave it as URL param.

It would help me if I understand what is your motivation, is it programatic access, or overall curl usage simplicity? Or are you suggesting that versioning of REST API should be handled differently?

Thanks for clarification, Emily.

BTW, do you have any take on it @rdebusscher  ?

yeekangc commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,248 @@
+
+    public static final String ERROR001 = "{\"error\":\"supportedServer query parameter is mandatory\"}";
+    public static final String ERROR002 = "{\"error\":\"Selected supportedServer is not available for given mpVersion\"}";
+    public static final String ERROR003 = "{\"error\":\"One or more selectedSpecs is not available for given mpVersion\"}";
```
Returning the error code too will be useful for programmatic access.

Karm replied:
Fixed.

cealsair commented:
Overall, great job!
One quick comment. When printing the help info on the command line as a response to:
curl https://test-start.microprofile.io/api/1.0.0

It would be nice to add a bit more explanation for the ETAG case. The explanation for the rest of the examples is pretty self-explanatory.

cealsair commented:
In README.md, please add a comment that points readers to the "how to use the command", etc. and link to the command help file. Thanks.

Karm commented on pom.xml:
```diff
@@ -37,7 +37,7 @@
 
     <properties>
-        <version.thorntail>2.2.1.Final</version.thorntail>
+        <version.thorntail>2.4.0.Final</version.thorntail>
```
Addressed.

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/API.java:
```diff
@@ -0,0 +1,77 @@
+    @Path("/project")
+    @GET
+    @Produces({"application/zip", "application/json"})
+    public Response getProject(@HeaderParam(HttpHeaders.IF_NONE_MATCH) String ifNoneMatch,
```
Addressed by taking app version from git into account.

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/API.java:
```diff
@@ -0,0 +1,77 @@
+    @POST
+    @Consumes({"application/json"})
+    @Produces({"application/zip", "application/json"})
+    public Response projectPost(@HeaderParam(HttpHeaders.IF_NONE_MATCH) String ifNoneMatch, @NotNull Project body) {
```
Addressed by taking app version from git into account.

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,248 @@
+/**
+ * @author Michal Karm Babacek <karm@redhat.com>
+ */
+@RequestScoped
```
Fixed,

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    public static final String ERROR001 = "{\"error\":\"supportedServer query parameter is mandatory\"}";
```
Uh, I use them in rest assured test I started, but I don't want to block this PR on CQ-ing Rest Assured.

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,248 @@
+            readme = s.hasNext() ? s.next() : "";
+            readmeEtag = new EntityTag(Integer.toHexString(readme.hashCode()));
+        } catch (Exception e) {
+            e.printStackTrace();
```
Fixed.

Karm commented:
@rdebusscher , @cealsair , @yeekangc 

Hi guys, the PR is ready for review, I think I addressed all suggestions. Feel free to fire new ones :smiley:

Karm commented:
(note that CloudFlare changes the etag, so it is different from what you get locally)

rdebusscher commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,251 @@
+        return engineData;
+    }
+
+    private Response processProject(String ifNoneMatch, @NotNull Project p) {
```
The @NotNull does nothing here (since it is just called as method) but more important, the @Pattern on groupId and artifactId are not checked/enforced.

Thus requesting a download with `$demo` as artifactId result in a non compilable project.

You can perform programmatic validation by

     Validator validator = Validation.buildDefaultValidatorFactory().getValidator();

The Validator retrieval can be in the init().

    Set<ConstraintViolation<Project>> violations = validator.validate(p);

Validates the Project.

PS. Make sure the - is also allowed in artifactId, see for example org.eclipse.microprofile.starter.validation.PackageValidator

Karm replied:
Hi, @rdebusscher,

I addressed the validation in a separate commit: 5b56762851914af4062e3116402b775b0d65e0c1 

## Intended changes
* changes validation pattern for artifactId and groupId for Web and REST 
* validation pattern is chnaged to  ```^(?:\w+|\w+[\.-]\w+)+$```  which removes the necessity to swap minuses with dots
* JSF reports the name of the field now
* maximum length is set to 100 for artifactId and the same for groupId
* REST API shares the same validation pattern

## Web changes
Added field name:
![validation1](https://user-images.githubusercontent.com/691097/59074397-81448200-88cb-11e9-8dcd-612f80cf57d4.png)

## Validator, ConstraintViolations
I was looking into adding annotations to the POJO and then using ```validator = Validation.buildDefaultValidatorFactory().getValidator()``` in postConstruct and getting violations in my validate method with ```validator.validate(p)```, but I decided against it as it splits validation to several places.

The current implementation does the basic sanity on deserialization, e.g. when user supplies non-existing Enum. Then all the rest of the validation is done in validate method, including setting default values where permissible.

if you don't like it, I can change it again. if it is the case, I would appreciate some hint on how to do it cleanly. Perhaps ditching validate method and creating validator and some "defaults setter". Not sure.

Karm replied:
@rdebusscher BTW, I cannot click on that red cross :x: to close the message. Firefox 66.0.5 Linux x86_64. Not sure if this is our fault or a bug for Primefaces...

rdebusscher replied:
Not a bug, it is just an icon indicating the severity of the message. But I agree, confusing. See #161

rdebusscher commented on src/main/java/org/eclipse/microprofile/starter/core/validation/PackageNameValidator.java:
```diff
@@ -23,17 +23,16 @@
 
-    private static final String VALID_REGEX = "^(?:\\w+|\\w+\\.\\w+)+$";
+    public static final String VALID_REGEX = "^(?:\\w+|\\w+[\\.-]\\w+)+$";
+    public static final int MAX_LENGTH = 100;
```
@Karm Any reason for the max length?

Karm replied:
@rdebusscher While I was looking at the validation, it seemed to me we should cap this at some limit, otherwise the limit is server's default which is something like 10MB per request (headers+body) for Thorntail (Undertow) IIRC. I am much more a httpd guy than a Java/Rest Easy one though, so there might be much more elegant way of doing this...

rdebusscher commented on src/main/java/org/eclipse/microprofile/starter/validation/PackageValidator.java:
```diff
@@ -41,10 +41,9 @@
+        if (!validator.isValidPackageName(value.toString())) {
             FacesMessage msg =
-                    new FacesMessage("Field validation failed.",
+                    new FacesMessage(uiComponent.getId() + " field validation failed.",
```
id should be internal and not necessary reflecting the label of the field (but in our case, id is the same as the label used for the field)

Better is to use label property.

java: `((HtmlInputText)uiComponent).getLabel()`
xhtml: add `label="groupId"`  on the <p:inputText>

Karm commented:
Hi @rdebusscher , thx for the feedback. Is it all for this iteration? If yes, I'll fix the license headers and the labels then...

rdebusscher commented:
@Karm No more remarks from more side, maybe Emily has some more things ...

Maybe change 100 to 200 than if you want to define a limit.

Karm commented:
@rdebusscher  See cb091fb

@Emily-Jiang If you don't have anything else, let's merge :smiley:

Emily-Jiang commented on src/main/resources/REST-README.md:
```diff
@@ -0,0 +1,246 @@
+    "PAYARA_MICRO",
+    "LIBERTY",
+    "TOMEE",
+    "KUMULUZEE"
```
I suggest order the list alphabetically across the board.

Karm replied:
Hello @Emily-Jiang, thank you for the feedback. Let me comment below:

> Why THORNTAIL is not listed? Do we expect the latest version is used for all runtime?

[Thorntail is not listed as supporting MP14 in this application](https://github.com/Karm/microprofile-starter/blob/rest/src/main/java/org/eclipse/microprofile/starter/addon/microprofile/servers/model/SupportedServer.java#L35)

> Will you have to update THORNTAIL image name when THORNTAIL_3 is released?

Whatever is in this [Enum](https://github.com/Karm/microprofile-starter/blob/rest/src/main/java/org/eclipse/microprofile/starter/addon/microprofile/servers/model/SupportedServer.java#L31) appears in both Web and REST. It is not the point of this PR to change it at this point I would say. This is not REST specific - we use it for Web too and we have been doing so from the first Beta launch...

> I suggest order the list alphabetically across the board.

The order of servers is randomized on Web to avoid bias...I did the same for the REST API on purpose. I would like us to be consistent, so if we have REST in lexicographical order, we should do it for the Web too. I suggest a separate Issue and a PR discussion for that.

Emily-Jiang replied:
ok

Emily-Jiang commented:
One general comment: I suggest the endpoint should always be the latest release of the starter. Therefore, 1.0.0 should be removed from the url.

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,297 @@
+    public static final String ERROR001 =
+            "{\"error\":\"supportedServer query parameter is mandatory\",\"code\":\"ERROR001\"}";
+    public static final String ERROR002 =
+            "{\"error\":\"Selected supportedServer is not available for given mpVersion\",\"code\":\"ERROR002\"}";
```
Addressed.

Karm commented on src/main/java/org/eclipse/microprofile/starter/rest/APIService.java:
```diff
@@ -0,0 +1,297 @@
+    public static final String ERROR002 =
+            "{\"error\":\"Selected supportedServer is not available for given mpVersion\",\"code\":\"ERROR002\"}";
+    public static final String ERROR003 =
+            "{\"error\":\"One or more selectedSpecs is not available for given mpVersion\",\"code\":\"ERROR003\"}";
```
Addressed.

Karm commented on src/main/resources/REST-README.md:
```diff
@@ -0,0 +1,246 @@
+    "CONFIG",
+    "FAULT_TOLERANCE",
+    "JWT_AUTH",
+    "HEALTH_METRICS",
```
@Emily-Jiang I cannot do that. I aimed for no code duplication, so as one can seamlessly add specs and servers in one place and these are used both for Web and REST without any additional code changes.  i.e. I did not come up with these constants, they have been there all along, see [MicroprofileSpec.java](https://github.com/Karm/microprofile-starter/blob/rest/src/main/java/org/eclipse/microprofile/starter/addon/microprofile/servers/model/MicroprofileSpec.java#L41)

rdebusscher replied:
We can change these 'codes' like HEALTH_METRICS, if we also change the variables in the templates. But I would do this in another PR.

Emily-Jiang replied:
ok

Karm commented:
@Emily-Jiang @rdebusscher 

So, there seems to be a concern about ```/api/1.0.0/```.  How about we keep it there as it is, since we are likely to be changing stuff and we would want API versions (perhaps shortened to  ```/api/1/```)
AND we add a shim (preferably in a separate PR) that would detect ```User-Agent``` and redirect tot he REST endpoint servlet? e.g. if ```User-Agent``` contains *curl* or *PowerShell* or *wget* we redirect from the landing page to the REST endpoint - the latest one.

rdebusscher commented:
@Emily-Jiang @Karm 
I do like the explicit versioning, especially because this endpoint will be used by tools.
We can foresee that there is an 'unversioned' one which always point to the latest. (just another JAX-RS resource which internally calls the latest code)

I don't like the fact that we have to look at the user-agent to know what we need to do (Web or REST) and complicates it unnecessarily.. So /api (or whatever name) as prefix is preferred.

Karm commented:
@Emily-Jiang @rdebusscher 

Based on your feedback, I did this:

* /api/ path param stays

* our first API version is called *1* instead of *1.0.0* to address Emily's concern about URL

* we have a default */* that maps to latest endpoint; which is *1* in our case

* it is easy to extend and override to get version 2, e.g.  adding:
```
@Path("/2")
public class APIEndpointV2 extends APIEndpointLatest {
    @Path("/")
    @GET
    @Produces({"text/x-markdown"})
    @Override
    public Response readme(@HeaderParam(HttpHeaders.IF_NONE_MATCH) String ifNoneMatch) {
        return Response.ok("LOL V2").build();
    }
}
```
gives us:

```
 curl 'localhost:8080/api/1'
Usual README output...

 curl 'localhost:8080/api/2'
LOL V2
```

Note "LOL V2" string in version 2 that overrides {{readme}} method.

You can try it live on curl https://test-start.microprofile.io/api
  
e.g. 

 * ```curl https://test-start.microprofile.io/api/1```
 * ```curl -O -J https://test-start.microprofile.io/api/project?supportedServer=PAYARA_MICRO```
 * ```curl -O -J https://test-start.microprofile.io/api/1/project?supportedServer=PAYARA_MICRO```

etc.


I hope we can move this PR on and address new things in separate PRs as we go.

yeekangc commented:
What we have now is Maven only, correct?  What is the thinking for the API when say Gradle will be supported?

A new path or another parameter to be provided?  I feel it should be former but not sure if this has been discussed?

Karm commented:
@yeekangc It should be a parameter as it IMHO falls into the Java selection category of runtimes. JDK version is consumed as a parameter (defaults to 1.8) in the current implementation.

Karm commented:
I have acks both from @Emily-Jiang and @rdebusscher 

Merging

yeekangc commented:
Thank you, @Karm.  I wonder if it should be something like "/project/maven" and "/project/gradle" but we can discuss that when Gradle support is added.
